cpp/src/arrow/array/builder_dict.cc

```diff
@@ -30,11 +33,117 @@
 
 namespace arrow {
 
 using internal::checked_cast;
 
+// ----------------------------------------------------------------------
+// DictionaryType unification
+
+struct UnifyDictionaryValues {
+  MemoryPool* pool_;
+  std::shared_ptr<DataType> value_type_;
+  const std::vector<const DictionaryType*>& types_;
+  std::shared_ptr<Array>* out_values_;
+  std::vector<std::vector<int32_t>>* out_transpose_maps_;
+
+  Status Visit(const DataType&, void* = nullptr) {
+    // Default implementation for non-dictionary-supported datatypes
+    std::stringstream ss;
+    ss << "Unification of " << value_type_->ToString()
+       << " dictionaries is not implemented";
+    return Status::NotImplemented(ss.str());
+  }
+
+  template <typename T>
+  Status Visit(const T&,
+               typename internal::DictionaryTraits<T>::MemoTableType* = nullptr) {
+    using ArrayType = typename TypeTraits<T>::ArrayType;
+    using DictTraits = typename internal::DictionaryTraits<T>;
+    using MemoTableType = typename DictTraits::MemoTableType;
+
+    MemoTableType memo_table;
+    if (out_transpose_maps_ != nullptr) {
+      out_transpose_maps_->clear();
+      out_transpose_maps_->reserve(types_.size());
+    }
+    // Build up the unified dictionary values and the transpose maps
+    for (const auto& type : types_) {
+      const ArrayType& values = checked_cast<const ArrayType&>(*type->dictionary());
+      if (out_transpose_maps_ != nullptr) {
+        std::vector<int32_t> transpose_map;
+        transpose_map.reserve(values.length());
+        for (int64_t i = 0; i < values.length(); ++i) {
+          int32_t dict_index = memo_table.GetOrInsert(values.GetView(i));
+          transpose_map.push_back(dict_index);
+        }
+        out_transpose_maps_->push_back(std::move(transpose_map));
+      } else {
+        for (int64_t i = 0; i < values.length(); ++i) {
+          memo_table.GetOrInsert(values.GetView(i));
+        }
+      }
+    }
+    // Build unified dictionary array
+    std::shared_ptr<ArrayData> data;
+    RETURN_NOT_OK(DictTraits::GetDictionaryArrayData(pool_, value_type_, memo_table,
+                                                     0 /* start_offset */, &data));
+    *out_values_ = MakeArray(data);
+    return Status::OK();
+  }
+};
+
+Status DictionaryType::Unify(MemoryPool* pool, const std::vector<const DataType*>& types,
+                             std::shared_ptr<DataType>* out_type,
+                             std::vector<std::vector<int32_t>>* out_transpose_maps) {
+  if (types.size() == 0) {
+    return Status::Invalid("need at least one input type");
+  }
+  std::vector<const DictionaryType*> dict_types;
+  dict_types.reserve(types.size());
+  for (const auto& type : types) {
+    if (type->id() != Type::DICTIONARY) {
+      return Status::TypeError("input types must be dictionary types");
+    }
+    dict_types.push_back(checked_cast<const DictionaryType*>(type));
+  }
+
+  // XXX Should we check the ordered flag?
```

Comment:
This is a bit tricky. We should consider it in a follow up JIRA.

You could conceivably have a sequence of dictionaries like

```
a < b < c
a < b < c < d
a < b < c < d < e < f
```

This could result from a sequence of ordered delta dictionaries. You could argue in this case that the output type should also be ordered.

Reply:
IMHO the "ordered" flag doesn't seem to make sense. A dictionary is simply a linear lookup table, so it's always ordered in a sense.

Reply:
ordered is part of the protocol https://github.com/apache/arrow/blob/master/format/Schema.fbs#L236

Reply:
Minor optimization: if dictionaries have an ordering, that could be used to merge them without building a memo table.

Reply:
@bkietz I'm not sure I understand your suggestion. "ordered" here does not mean "sorted", just that the dictionary's order is semantically meaningful.

Reply:
@pitrou My understanding is that if dictionaries to be unified are ordered, then the unified dictionary must preserve the order of shared emements. This in turn requires that those shared elements be identically ordered in each of the input dictionaries. For example, it seems impossible to unify `[0, 1], ordered` and `[1, 0], ordered` since the first claims it is semantically significant that 0 precedes 1 while the second claims the opposite. If shared elements are identically ordered then the new dictionary can be derived using longest common subsequence

Reply:
```
A   0     1 2 3 4 5       6

B     7 8 1   3 4   10 11 6

U   0 7 8 1 2 3 4 5 10 11 6
```

Reply:
Right, but how does the algorithm look like and why would it more efficient than a series of hash lookups?

Reply:
Sorry, you're right; talking about optimization is not justified when there is not a virgin which preserves order using a hash table. I'll write an example of each and benchmark them

Reply:
I would say it's not worth spending time right now on handling multiple "ordered" dictionaries (where the only case where "ordered" would be preserved is when all the dictionaries are prefixes of a common dictionary)

Reply:
I created https://issues.apache.org/jira/browse/ARROW-4096. This is low priority